libslide/src/evaluator_rules/pattern_matching.rs

```diff
@@ -122,78 +152,109 @@ impl Replacements {
+            (Stmt::Expr(expr), _) => expr,
+            _ => unreachable!(),
+        }
+    }
 
     mod replacements {
         use super::*;
 
         #[test]
         fn try_merge() {
-            let a = ExprPat::VarPat("a".into());
-            let b = ExprPat::VarPat("b".into());
-            let c = ExprPat::VarPat("c".into());
+            let a = Rc::new(ExprPat::VarPat("a".into()));
+            let b = Rc::new(ExprPat::VarPat("b".into()));
+            let c = Rc::new(ExprPat::VarPat("c".into()));
 
             let mut left = Replacements::default();
-            left.insert(a.clone(), Expr::Const(1.));
-            left.insert(b.clone(), Expr::Const(2.));
+            left.insert(&a, Expr::Const(1.).into());
+            left.insert(&b, Expr::Const(2.).into());
 
             let mut right = Replacements::default();
-            right.insert(b.clone(), Expr::Const(2.));
-            right.insert(c.clone(), Expr::Const(3.));
+            right.insert(&b, Expr::Const(2.).into());
+            right.insert(&c, Expr::Const(3.).into());
 
             let merged = Replacements::try_merge(left, right).unwrap();
             assert_eq!(merged.map.len(), 3);
-            assert_eq!(merged.map.get(&a).unwrap().to_string(), "1");
-            assert_eq!(merged.map.get(&b).unwrap().to_string(), "2");
-            assert_eq!(merged.map.get(&c).unwrap().to_string(), "3");
+            assert_eq!(merged.map.get(&hash(&a)).unwrap().to_string(), "1");
+            assert_eq!(merged.map.get(&hash(&b)).unwrap().to_string(), "2");
+            assert_eq!(merged.map.get(&hash(&c)).unwrap().to_string(), "3");
         }
 
         #[test]
         fn try_merge_overlapping_non_matching() {
-            let a = ExprPat::VarPat("a".into());
+            let a = Rc::new(ExprPat::VarPat("a".into()));
 
             let mut left = Replacements::default();
-            left.insert(a.clone(), Expr::Const(1.));
+            left.insert(&a, Expr::Const(1.).into());
 
             let mut right = Replacements::default();
-            right.insert(a, Expr::Const(2.));
+            right.insert(&a, Expr::Const(2.).into());
 
             let merged = Replacements::try_merge(left, right);
             assert!(merged.is_none());
         }
+
+        #[test]
+        fn transform_common_subexpression_elimination() {
+            let parsed_rule_from = parse_rule("#a * _b + #a * _b");
+            let parsed_rule_to = parse_rule("#a * _b + #a * _b");
+            let parsed_target = parse_expr("0 * 0 + 0 * 0");
+
+            let repls = match_rule(Rc::new(parsed_rule_from), Rc::new(parsed_target)).unwrap();
+            let transformed = repls.transform(Rc::new(parsed_rule_to));
+            let (l, r) = match transformed.as_ref() {
+                Expr::BinaryExpr(BinaryExpr { lhs, rhs, .. }) => (lhs, rhs),
+                _ => unreachable!(),
+            };
+            assert!(std::ptr::eq(l.as_ref(), r.as_ref())); // #a * _b
+
```

Comment:
again in 223-225 youre just pulling the enum variant out. Why use a match?

Reply:
See a couple of comments above.

1. To pull an enum variant out we need to perform pattern matching anyway (there is no other way save memory discrimination, which is << pattern matching IMO)

2. We are then pulling out a `BinaryExpr` held by the variant, which also requires matching.